### ISSUE TITLE
Добавить возможность "архивировать" контакт.

### ISSUE THREAD

**Issue by SM-Sergey** [miranda-ng/miranda-ng#3999]
Задача. Имеются десятки контактов, которые де-факто давно "умерли", но висят в серверном клисте разных протоколов только ради того, чтобы можно было полистать историю общения по ним за разное прошлое время, включая и то, когда какие-то протоколы вообще не имели истории на сервере. Надо их поудалять из серверных контакт-листов, чтобы их "почистить", и локально куда-то перенести в какую-то группу контактов, отвязав от протоколов, оставив только историю, и "следы" протокола - откуда этот контакт произрос.
На форуме ответа, как это сделать, не дали. Поэтому, открываю тикет. 

Я вижу эту реализацию такой:

- По ПКМ на контакте в клисте должно появиться меню "Archive contact".
- По его нажатию должно произойти следующее: протокол должен удалить контакт из клиста на сервере, если протокол это поддерживает, если это группа - протокол должен сделать так, чтобы мы покинули эту группу, если протокол это поддерживает; cам контакт и его история в БД должны перейти под управление "протокола" Dummy, который позволяет смотреть историю, а текущий протокол забыть об этом контакте; Контакт должен улететь в отдельную группу контактов например "Archived", ну можно настраивать в какую; Информация о том, какой был до этого протокол, и все его настройки, связанные с этим протоколом, должны сохраниться в БД. Это вот главное.
- Желательно (не обязательно) иметь возможность восстановить историю этого контакта в протоколе. Ведь сто лет назад покинувший нас контакт может реинкарнироваться, в том числе и в другом обличии. И может понадобиться обратно вернуться в давно покинутую группу. Для этого надо возможность подтянуть историю из архивного контакта в какой-то новый. То есть, на любом контакте, по ПКМ на нем, надо иметь возможность "Add history from archive" -  иметь возможность "подгрузить историю из архивного контакта". По этому меню должны показаться все архивные контакты, которые были сархивированы из этого протокола, чтобы выбрать, откуда подтянуть историю.
- Архивный контакт можно только удалить уже совсем, необратимо.

Как другой вариант-  реализация может быть проще. Но требовать больше действий пользователя. Зато и возможностей больше.
- По "Archive contact" просто создать новый контакт с тем же видимым именем, в архивной группе, под управлением протокола "Dummy", и скопировать туда всё и вся из этого контакта, историю, старый протокол, настройки с ним связанные.
- Убить оригинальный контакт теперь можно и руками, для этотого и так есть "Delete" или "Leave chat". А можно и не убивать, если не хочется.

**Comment by SM-Sergey** [comment]
Да, пожалуй второй вариант туп, прост и эффективен. И не надо ничего больше. Просто надо создать копию контакта с историей и информацией о старом протоколе, но под управлением dummy. И без мудростей. И, может, в каком-то далеком будущем, возможность подтянуть историю из него обратно в живой контакт того же протокола.

**Comment by dartraiden** [comment]
Пока я вижу это так
- проверяем, есть ли учётка Dummy с внутренним именем "ARCHIVE", если её нет создаём
- делаем копию контакта с заменой значения ключа `/Protocol/p` на "ARCHIVE" и имени модуля протокола тоже на "ARCHIVE"
- мочим исходный контакт, чтобы протокол синхронизировал удаление с серверным ростером

**Comment by SM-Sergey** [comment]
И вот еще. Если это будет реализовано, то логично спрашивать при удалении контакта или покидании группы, хотите ли Вы его сархивировать, или прибить насмерть.

**Comment by georgehazan** [comment]
Пара мыслей к вопросу

1. Гораздо проще не копировать контакт (зачем?), а просто перебить у него учетку на Dummy.
2. В миранде нет возможности  в одной учетке сделать так, чтобы контакты имели unique id разного типа и имени. Резюме - возможно, для каждого из протоколов придется делать свой архив

**Comment by SM-Sergey** [comment]
1. Для того, чтобы иметь архивную копию живого контакта, с которым в основной записи могут любые чудеса случиться, как то прилететь с сервера ошибочные (или не ошибочные) стирания, или миранда сглючить и что-то грохнуть зазря. А тут раз - и есть архивная копия контакта со всей историей. Можно, конечно, держать кучу архивных копий всей БД, но логично иметь архивные копии только наиболее важных контактов с историями. Ну и да, если архивирование контакта сопряжено с его удалением из протокола - то да, почему бы и не просто перебить его на Dummy.
2. А сохранять ID в общем, и незачем. Его логично сделать новым. Главное - сохранить на всякий случай UID/логин/пароль от протокола. Ну, и если это для чего-то может понадобиться, можно отдельно сохранить и оригинальный ID.

**Comment by dartraiden** [comment]
Я за перебить. Но сейчас, когда перебиваешь руками, контакт превращается в "неизвестный контакт" до перезапуска. Этого быть не должно.

**Comment by SM-Sergey** [comment]
> Я за перебить.

Поясните пожалуйста, как при этом реализовать архивирование действующего контакта без его удаления из протокола? Или это подразумевает два метода - для контактов без удаления копирование, а для контактов с удалением - перебивание?

**Comment by dartraiden** [comment]
Очень просто: контакту перебивается протокол, при следующем выходе в сеть контакт снова прилетает с сервера.

**Comment by dartraiden** [comment]
ToDo: придумать какой-то обходной путь для https://github.com/miranda-ng/miranda-ng/issues/4359

**Comment by SM-Sergey** [comment]
> либо вгружает историю с сервера (если это ему нужно).

Так это не сработает, так как существует очень много истории, десятки тысяч сообщений у контактов со времен, когда история на сервере еще не хранилась. Да и никакой гарантии нет, что на сервере все сохранилось и со старых времен, когда уже была серверная история, даже наоборот, очень часто на сервере много старого нету.

**Comment by dartraiden** [comment]
Пинг @georgehazan
